### PR TITLE
 plugins/telescope.undo: init

### PR DESCRIPTION
**Pull request by hmajid2301**
Added telescope extension: [telescope-undo](https://github.com/debugloop/telescope-undo.nvim)

### CODE REVIEW

File: plugins/telescope/undo.nix [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+          helpers.ifNonNull' i
+          (mapAttrs
+            (
+              key: action:
+                helpers.mkRaw "require('telescope-undo.actions').${action}"
+            )
+            mappings);
+
+        n =
+          helpers.ifNonNull' n
```
Comment (hmajid2301) [inline plugins/telescope/undo.nix:121]: also not sure if referencing `n` here will cause issues with `shadowing` let me know :) (and same for the i)

File: plugins/telescope/undo.nix [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+    timeFormat = mkOption {
+      type = types.nullOr types.str;
+      description = ''
+        Can be set to a [Lua date format string](https://www.lua.org/pil/22.1.html).
+      '';
+      default = "";
+    };
+
+    mappings = {
```
[the remark this answers is not in the document]
Comment (GaetanLepage) [reply plugins/telescope/undo.nix:69]: That looks a perfectly valid solution to me !

File: plugins/telescope/undo.nix (hunk 1 above)
Comment (GaetanLepage) [reply plugins/telescope/undo.nix:121]: `shadowing` ?

**Review by GaetanLepage — CHANGES_REQUESTED** [review]

File: plugins/telescope/undo.nix (hunk 1 above)
Comment (hmajid2301) [reply plugins/telescope/undo.nix:121]: "reusing" the name var `n` twice ?

Comment (GaetanLepage) [reply plugins/telescope/undo.nix:121]: No, this should work just fine.
Could you please add some test in `tests/test-sources/plugins/telescope/undo.nix` please ?

File: plugins/telescope/undo.nix [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        )
+        ''
+           {
+                 "y" = "yank_additions";
+                 "Y" = "yank_deletions";
+                 "u" = "restore";
+          }
+        '';
+    };
+  };
```
Comment (GaetanLepage) [inline plugins/telescope/undo.nix:97]: Warning, you missed the description.
```suggestion
      i =
        helpers.defaultNullOpts.mkNullable
        (
          with types;
            attrsOf str
        )
        ''
          {
               "<cr>" = "yank_additions";
               "<s-cr>" = "yank_deletions";
               "<c-cr>" = "restore";
          }
        ''
        "<INSERT_DESCRIPTION>";
      n =
        helpers.defaultNullOpts.mkNullable
        (
          with types;
            attrsOf str
        )
        ''
           {
                 "y" = "yank_additions";
                 "Y" = "yank_deletions";
                 "u" = "restore";
          }
        ''
        "<INSERT_DESCRIPTION>";
    };
  };
```

File: plugins/telescope/undo.nix [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    diffContextLines = mkOption {
+      type =
+        helpers.defaultNullOpts.mkNullable
+        (
+          with types;
+            either ints.unsigned str
+        );
+      description = ''Defaults to the scrolloff'';
+      default = "vim.o.scrolloff";
+    };
```
Comment (GaetanLepage) [inline plugins/telescope/undo.nix:53]:
```suggestion
    diffContextLines = mkOption {
      type = with types;
        either ints.unsigned str;
      description = ''Defaults to the scrolloff'';
      default = "vim.o.scrolloff";
    };
```

**Review by GaetanLepage — CHANGES_REQUESTED** [review]
Why did you use `mkOption` instead of the `null`-defaulting options that we usually use ?

**Comment by hmajid2301** [comment]
> Why did you use `mkOption` instead of the `null`-defaulting options that we usually use ?

I used `fzf-native` plugin as an example and they used mkOption there. There also didnt seem to be any tests for these, hence I didn't create any for these :sweat:! 

But will add some tests now!!! :)

File: tests/test-sources/plugins/telescope/undo.nix [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+      extensions.undo = {
+        enable = true;
+
+        useDelta = true;
+        useCustomCommand = ["bash" "-c" "echo '$DIFF' | delta"];
+        sideBySide = true;
+        diffContextLines = 8;
+        entryFormat = "state #$ID";
+        timeFormat = "!%Y-%m-%dT%TZ";
+        mappings = {
```
Comment (hmajid2301) [inline tests/test-sources/plugins/telescope/undo.nix:19]: I think the test is failing here:

```bash
error:
       … while checking flake output 'checks'

         at /nix/store/p58d2j0ac7zvja5jl14xzbc19fakjxh2-source/lib.nix:39:17:

           38|               {
           39|                 ${key} = (attrs.${key} or { })
             |                 ^
           40|                   // { ${system} = ret.${key}; };

       … while checking the derivation 'checks.x86_64-linux.plugins-telescope-undo-example'

         at «none»:0: (source not available)

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: cannot coerce a set to a string

```

**Comment by GaetanLepage** [comment]
> I used `fzf-native` plugin as an example and they used mkOption there. There also didnt seem to be any tests for these, hence I didn't create any for these 😓!

Yes, this is totally fair. However, this is an example of an older part of the code base (untested + not much using the `lib/helpers.nix` functions).
It is better to switch to `helpers.defaultNullOpts.MkXxxx` where possible as well as testing.

File: plugins/telescope/undo.nix [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        n =
+          helpers.ifNonNull' n
+          (mapAttrs
+            (
+              key: action:
+                helpers.mkRaw "require('telescope-undo.actions').${action}"
+            )
+            mappings);
+      };
+    };
```
Comment (GaetanLepage) [inline plugins/telescope/undo.nix:127]:
```suggestion
      mappings = with mappings; {
        i =
          helpers.ifNonNull' i
          (
            mapAttrs
            (key: action:
              helpers.mkRaw "require('telescope-undo.actions').${action}")
            i
          );

        n =
          helpers.ifNonNull' n
          (
            mapAttrs
            (key: action:
              helpers.mkRaw "require('telescope-undo.actions').${action}")
            n
          );
      };
    };
```

File: tests/test-sources/plugins/telescope/undo.nix (hunk 5 above)
Comment (GaetanLepage) [reply tests/test-sources/plugins/telescope/undo.nix:19]: Tests now pass

File: plugins/telescope/undo.nix (hunk 6 above)
Comment (GaetanLepage) [reply plugins/telescope/undo.nix]: This still needs to be fixed.

**Review by GaetanLepage — APPROVED** [review]
Good job !

**Pull request merged**